porick/views.py

```diff
@@ -189,7 +190,69 @@ def logout():
     flash('Logged out successfully!', 'info')
     return response
 
-
-@app.route('/reset_password')
+@app.route('/reset_password', methods=['GET', 'POST'])
 def reset_password():
-    raise NotImplementedError()
+    if request.method == 'GET':
+        if not request.args.get('key'):
+            return render_template('/pw_reset/request.html')
+
+        token = PasswordReset.query.filter_by(key=request.args['key']).first()
+        if token and token.is_valid:
+            return render_template('/pw_reset/set.html', key=token.key)
+        else:
+            flash('Invalid reset token', 'error')
+            return render_template('/index.html')
+
+
+    elif request.method == 'POST':
+        if not request.args.get('key'):
+        # The user has requested a new token
+            email = request.form['email']
+            user = User.query.filter_by(email=email).first()
+            if not user:
+                flash('Invalid email address provided.', 'error')
+                return render_template('/pw_reset/request.html')
+
+            # If this user already has a valid reset token, don't
+            # let them create a new one until it has expired
+            existing_token = PasswordReset.query.filter_by(user_id=user.id).first()
+            if existing_token:
+                if existing_token.is_valid:
+                    flash('A password reset has already been requested for this user.', 'error')
+                    return render_template('/pw_reset/request.html')
+                else:
+                    db.session.delete(existing_token)
+                    db.session.commit()
+
+            token = PasswordReset(user.id)
+            key = token.key
+            db.session.add(token)
+            db.session.commit()
+            send_reset_password_email(email, key)
+            flash('Password reset email sent!', 'success')
+            return render_template('/index.html')
+
+        else:
+        # Reset the password to what they provided
+            token = PasswordReset.query.filter_by(key=request.args['key']).first()
+            if not token or not token.is_valid:
+                flash('Invalid reset token.', 'error')
+                return render_template('/pw_reset/request.html')
+
+            password = request.form['password']
+            confirm_password = request.form['password_confirm']
+            valid_password = validate_password(password, confirm_password)
+            if valid_password['status']:
+                user = User.query.get(token.user_id)
+                user.password = hash_password(password)
+                db.session.add(user)
```

Comment:
I believe this will fail - I don't think there's any need to do an `add()`. Once you've changed the attribute's value, you can just do a `commit()` and it's persisted to the db.